blog/2021-06-14-stryker4s-sbt-performance.mdx

```diff
@@ -0,0 +1,151 @@
+
+In this blog we'll talk about the performance improvements made in the last couple Stryker4s versions and dive into how they are made possible.
+
+<!--truncate-->
+
+If you're new to mutation testing, it's a way to measure your tests' effectiveness. A mutation testing framework will make small changes, called _mutants_, one by one in your source code. Then it will run your tests to see if one of them fails. If so, you just "killed" that mutant; if not, it "survived". If too many mutants survive, you probably want to improve your tests. The mutation testing report will give you insides into the test cases you may have missed. If this all sounds complicated, please take a look at the recent talk ["Who is testing your tests?"](https://www.youtube.com/watch?v=Vq9eqZzblfg) from _Scala Love in the City_ or check out the rest of the [website](/).
+
+Imagine you're on your way to a sunny holiday destination. You've checked in your luggage, gone through customs, and boarded the plane. Then, the plane moves forward a couple of meters, and you're thrown out and sent back to customs before going through the whole thing again. It'll take years before you get to sip on your umbrella cocktail 🍹! This is exactly how Stryker4s has been running your code… until now.
+
+## 📃 In short
+
+A lot has changed in recent versions of Stryker4s. Stryker4s now runs your tests in an entirely different way, bringing big performance improvements. As an example, running Stryker4s on the Stryker4s codebase used to take about 40 minutes on `0.9.1`. On`0.12.1`, this has been reduced to a fast 40 seconds. A short list of some new features:
+
+- ✨ New testrunner boasting improved performance
+- 🔀 Concurrency for even faster performance
+- 🕵️‍♀️ Coverage analysis to skip mutants without coverage
+- ⏱️ Timeout detection in tests
+- 👽 New mutator: regular expressions. Thanks to [Weapon regeX](https://github.com/stryker-mutator/weapon-regex/)!
+
+In the rest of this blog post, we'll do a deep-dive into the increased performance. Let's start by taking a look at the older version of Stryker4s.
+
+## 🐌 The old
+
+In the older sbt plugin, there was a lot of overhead in running your tests. As a quick reminder, Stryker4s uses [mutation switching](https://stryker-mutator.io/blog/mutation-switching/), so it only has to compile your code once, instead of for each mutant. After mutating your source code, Stryker4s would essentially just call `sbt test` with the mutated code for every mutant. This means sbt would:
+
+1. look for code changes (which there never are).
+1. make sure all compiled code is up-to-date (which it always is).
+1. Start a new Java process with your class files, dependencies, and a small sbt entry point library.
+1. Run your tests and report the results back to the main sbt process.
+1. Repeat for the next mutant.
+
+import Mermaid from '@theme/Mermaid';
+
+<Mermaid
+  _key="sbt-old"
+  chart={`
+sequenceDiagram
```

Comment:
Suggestion:

```
sequenceDiagram
  participant Stryker4s
  participant sbt
  participant sbt test process
  Stryker4s->>+sbt: Compile mutated code
  sbt-->>-Stryker4s: bytecode
  loop For each mutant
    Stryker4s->>+sbt: executeTests
    sbt->>sbt: Check for changes?
    sbt->>+sbt test process: Start new process and run tests
    sbt test process-->>-sbt: Report test results
    sbt-->>-Stryker4s: Tests successful Y/N
  end
```

(sorry, don't know how to do multiline suggestions 🙈)

Reply:
This is how I had it first. It makes sense for this image, but then sbt and Stryker4s are swapped in the diagram below. What do you think is best?